### PR TITLE
Implement listmap keys uniqueness

### PR DESCRIPTION
**Pull request by yongruilin**
- Added UniqueByFunc to validate uniqueness of list items using a custom comparison function.
- Introduced test cases for UniqueByFunc in each_test.go to ensure correct functionality.
- Updated existing validation logic to utilize UniqueByFunc for various struct fields in generated validation files.
- Enhanced the validation framework to support early tag validation for improved processing order.

ref: https://github.com/jpbetz/validation-gen/issues/62

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/api/validate/each_test.go [hunk 1]
```diff
@@ -222,3 +222,27 @@ func testUniqueByReflect[T any](t *testing.T, name string, input []T, wantErrs i
...
 }
+
+type listItem struct {
+	key1 string
+	key2 string
+	val  string
+}
+
+func TestUniqueByFunc(t *testing.T) {
+	testUniqueByFunc(t, "valid", []listItem{{"key1", "key2", "val1"}, {"key1", "key2", "val2"}, {"key1", "key3", "val3"}}, 1)
```
Comment (aaron-prindle) [inline staging/src/k8s.io/apimachinery/pkg/api/validate/each_test.go:233]: nit: Currently there are duplicate tests with the name "valid", can make the name more descriptive to explain what specifically is being tested (that or drop the test name field)

File: staging/src/k8s.io/apimachinery/pkg/api/validate/each_test.go [hunk 2]
```diff
@@ -222,3 +222,27 @@ func testUniqueByReflect[T any](t *testing.T, name string, input []T, wantErrs i
...
+	testUniqueByFunc(t, "valid", []listItem{{"key1", "key2", "val1"}, {"key1", "key2", "val2"}, {"key1", "key2", "val3"}}, 2)
+	testUniqueByFunc(t, "valid", []listItem{{"key1", "key2", "val1"}, {"key3", "key4", "val2"}, {"key5", "key6", "val3"}}, 0)
```
Comment (aaron-prindle) [inline staging/src/k8s.io/apimachinery/pkg/api/validate/each_test.go:235]: Can we add tests for empty lists as well?  Single item lists might also make sense here.

File: staging/src/k8s.io/apimachinery/pkg/api/validate/each.go [hunk 3]
```diff
@@ -102,6 +102,13 @@ func UniqueByReflect[T any](_ context.Context, op operation.Operation, fldPath *
 	return unique(fldPath, newSlice, SemanticDeepEqual)
 }
 
+// UniqueByFunc verifies that each element in the list map is unique by using the provided comparison function.
+// The comparison function (cmp) determines when two elements are considered equal, and if any elements
+// are found to be equal according to this function, they are reported as duplicates.
+func UniqueByFunc[T any](_ context.Context, op operation.Operation, fldPath *field.Path, newSlice, _ []T, cmp CompareFunc[T]) field.ErrorList {
+	return unique(fldPath, newSlice, cmp)
```
Comment (aaron-prindle) [inline staging/src/k8s.io/apimachinery/pkg/api/validate/each.go:109]: Q: Is there any ratcheting consideration necessary for if a listMapKey=... is changed or an additional key is added to a k8s object over time? Not sure if this is allowable but might make sense to note in KEP or a comment somewhere that it was considered.

**Comment by thockin** [comment]
I have a sketch of how this can work, will try to post it today

**Review by thockin — COMMENTED** [review]
I came at this a different way but headed in the same direction.  Rather than make "early" to match "late", I defined "FieldValidators" which run on every field *after* all tags.

That means we would have TypeValidator, TagValidator, and FieldValidator which makes sense to me.

https://github.com/jpbetz/validation-gen/compare/main...thockin:validation-gen:vg_ensure_unique_listMap_keys?expand=1

Mine is WIP because I didn't like how it produces errors for non-unique keys -- it logs the whole type in `%#v` syntax (gross).  I was trying to make something that looks nicer, but got distracted.

Do you want to keep running with this one?

**Comment by thockin** [comment]
Here's my thinking, tell me if you disagree.

1) I hate `LateTagValidator` - it always felt like a hack.  Importantly if we do an Early version, it's possible for a validator to claim to be BOTH early and late.

2) We already have `TypeValidator` which runs on every type.  `FieldValidator` felt consistent.

3) The only existing user of `LateTagValidator` should probably be a `FieldValidator`.

I'm worried that we have built a dependency graph without putting a lot of thought into it.  Should we instead do something like this?
* each TagValidator has a method which returns a list of other tag names that it depends on
* in the sorting function, use that method to build a dependency-ordered list
* detect cycles and error-out

Pretty complicated, IMO.

Maybe we can write down rules about ordering:
- Tag validators always come first, for both types and fields.
- Type and Field validators always run after any Tag validators, and after the type or field has been processed.
- Tag validators MUST NOT depend on other Tag validators having run already.  Even when there are several tags that "work together", users might specify tags in the wrong order. Instead, they can accumulate information internally and use a Type or Field validator to finish the job.
- Type and Field validators MUST NOT depend on other Type or Field validators having run already.

**Comment by thockin** [comment]
I updated https://github.com/jpbetz/validation-gen/compare/main...thockin:validation-gen:vg_ensure_unique_listMap_keys?expand=1 a little - the last commit is starting to think about how to produce better errors, but not strictly required.

**Comment by thockin** [comment]
https://github.com/jpbetz/validation-gen/pull/97

Agree we can keep LateTagValidator for now and fix it when we have time - it's not urgent.  #95
